src/Interpreters/InterpreterCreateFunctionQuery.cpp

```diff
@@ -33,24 +38,90 @@ BlockIO InterpreterCreateFunctionQuery::execute()
     create_function_query.if_not_exists = false;
     create_function_query.or_replace = false;
+    create_function_query.is_temporary = false;
+
+    if (FunctionFactory::instance().hasNameOrAlias(function_name))
+    {
+        if (if_not_exists)
+            return {};
+
+        throw Exception(ErrorCodes::FUNCTION_ALREADY_EXISTS, "The function '{}' already exists", function_name);
+    }
+
+    if (AggregateFunctionFactory::instance().hasNameOrAlias(function_name))
+    {
+        if (if_not_exists)
+            return {};
+
+        throw Exception(ErrorCodes::FUNCTION_ALREADY_EXISTS, "The aggregate function '{}' already exists", function_name);
+    }
+
+    if (UserDefinedExecutableFunctionFactory::instance().has(function_name, current_context))
+    {
+        if (if_not_exists)
+            return {};
+
+        throw Exception(ErrorCodes::FUNCTION_ALREADY_EXISTS, "The user defined executable function '{}' already exists", function_name);
+    }
+
+    auto & global_user_defined_function_factory = UserDefinedSQLFunctionFactory::instance();
+
+    std::shared_ptr<UserDefinedSQLFunctionFactory> session_user_defined_function_factory;
+
+    if (current_context->hasSessionContext())
+        session_user_defined_function_factory = current_context->getSessionContext()->getUserDefinedSQLFunctionFactory();
 
     validateFunction(create_function_query.function_core, function_name);
-    user_defined_function_factory.registerFunction(current_context, function_name, query_ptr, replace, if_not_exists, persist_function);
+
+    UserDefinedSQLFunctionFactory::Lock global_factory_lock = global_user_defined_function_factory.getLock();
+    UserDefinedSQLFunctionFactory::Lock session_factory_lock;
+
+    if (session_user_defined_function_factory)
+        session_factory_lock = session_user_defined_function_factory->getLock();
+
+    bool function_exists_in_global_factory = global_user_defined_function_factory.has(global_factory_lock, function_name);
+    bool function_exists_in_session_factory = false;
+
+    if (session_user_defined_function_factory)
+        function_exists_in_session_factory = session_user_defined_function_factory->has(session_factory_lock, function_name);
+
+    bool function_already_exists = function_exists_in_global_factory || function_exists_in_session_factory;
+
+    if (!replace && function_already_exists)
+    {
+        if (if_not_exists)
+            return {};
+
+        throw Exception(ErrorCodes::FUNCTION_ALREADY_EXISTS, "Function with name '{}' already exists", function_name);
+    }
+
+    if (is_temporary)
+    {
+        if (!session_user_defined_function_factory)
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Session context should exist during CREATE TEMPORARY FUNCTION execution");
+
+        session_user_defined_function_factory->registerFunction(session_factory_lock, current_context, function_name, query_ptr, replace, false);
+        return {};
+    }
+
+    if (function_exists_in_session_factory)
```

Comment:
So, if temporary function exists, and we want to replace it (with non-temporary), then it would also be temporary, even if used did not specify it?

I think it is not obvious. It's better to either:
1. remove function from session factory and create it in global factory
2. throw exception

Also, as I understand, it is possible that function with the same name may exist in both factories?

Reply:
CREATE OR REPLACE will replace function temporary or non temporary, that will be find. There is no syntax for `CREATE OR REPLACE TEMPORARY`, need to check for tables, probably for `CREATE OR REPLACE TABLE` we will just replace temporary table.

Reply:
It is not possible for function to exists in both factories, there is synchronization between them to avoid such thing.

Reply:
Seems like it's possible (and it works in unexpected way):
```
session1 :) create temporary function f as (x) -> x
session2 :) create function f as (x) -> x + 1
session1 :) select f(1)

┌─plus(1, 1)─┐
│          2 │
└────────────┘

```